server/sonar-process-monitor/src/test/java/org/sonar/application/process/SQProcessTest.java

```diff
@@ -0,0 +1,333 @@
+
+      verify(listener).onProcessEvent(A_PROCESS_ID, ProcessEventListener.Type.OPERATIONAL);
+    }
+    verifyNoMoreInteractions(listener);
+  }
+
+  @Test
+  public void operational_event_is_sent_once() {
+    ProcessEventListener listener = mock(ProcessEventListener.class);
+    SQProcess underTest = SQProcess.builder(A_PROCESS_ID)
+      .addEventListener(listener)
+      .build();
+
+    try (TestProcess testProcess = new TestProcess()) {
+      underTest.start(() -> testProcess);
+      testProcess.operational = true;
+
+      underTest.refreshState();
+      verify(listener).onProcessEvent(A_PROCESS_ID, ProcessEventListener.Type.OPERATIONAL);
+
+      // second run
+      underTest.refreshState();
+      verifyNoMoreInteractions(listener);
+    }
+  }
+
+  @Test
+  public void send_event_when_process_requests_for_restart() {
+    ProcessEventListener listener = mock(ProcessEventListener.class);
+    SQProcess underTest = SQProcess.builder(A_PROCESS_ID)
+      .addEventListener(listener)
+      .build();
+
+    try (TestProcess testProcess = new TestProcess()) {
+      underTest.start(() -> testProcess);
+
+      testProcess.askedForRestart = true;
+      underTest.refreshState();
+
+      verify(listener).onProcessEvent(A_PROCESS_ID, ProcessEventListener.Type.ASK_FOR_RESTART);
+
+      // flag is reset so that next run does not trigger again the event
+      underTest.refreshState();
+      verifyNoMoreInteractions(listener);
+      assertThat(testProcess.askedForRestart).isFalse();
+    }
+  }
+
+  @Test
+  public void stopForcibly_stops_the_process_without_graceful_request_for_stop() {
+    SQProcess underTest = SQProcess.builder(A_PROCESS_ID).build();
+
+    try (TestProcess testProcess = new TestProcess()) {
+      underTest.start(() -> testProcess);
+
+      underTest.stopForcibly();
+      assertThat(underTest.getState()).isEqualTo(Lifecycle.State.STOPPED);
+      assertThat(testProcess.askedForStop).isFalse();
+      assertThat(testProcess.destroyedForcibly).isTrue();
+
+      // second execution of stopForcibly does nothing. It's still stopped.
+      underTest.stopForcibly();
+      assertThat(underTest.getState()).isEqualTo(Lifecycle.State.STOPPED);
+    }
+  }
+
+  @Test
+  public void process_stops_after_graceful_request_for_stop() throws Exception {
+    ProcessLifecycleListener listener = mock(ProcessLifecycleListener.class);
+    SQProcess underTest = SQProcess.builder(A_PROCESS_ID)
+      .addProcessLifecycleListener(listener)
+      .build();
+
+    try (TestProcess testProcess = new TestProcess()) {
+      underTest.start(() -> testProcess);
+
+      Thread stopperThread = new Thread(() -> underTest.stop(1, TimeUnit.HOURS));
+      stopperThread.start();
+
+      // thread is blocked until process stopped
+      assertThat(stopperThread.isAlive()).isTrue();
+
+      // wait for the stopper thread to ask graceful stop
+      while (underTest.getState() != Lifecycle.State.STOPPING) {
+        Thread.sleep(1L);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)